### PR TITLE
Allow non-UTF8 keys for decoding too

### LINKED ISSUE #2: `decode` doesn't take arbitrary bytes as input
I have a key that is not valid UTF-8. Currently it's possible to encode, but not decode using such a key.

I've been working on a PR that makes `decode` take `AsRef<[u8]>` (like `encode`), but on the way I've found that this breaks `decode::<Claims>(..)` usage because of additional type parameters. The other options are:

* `decode<T: Part>(token: &str, secret: &AsRef<[u8]>, algorithm: ...) -> Result<T, ...>`, no idea why this works???

* Just advise users to use different forms of typehinting if possible, and only use `decode::<Claims, _>` if necessary.

### PR DESCRIPTION
**Pull request by untitaker**
Fix #2

### CODE REVIEW

**Comment by untitaker** [comment]
I just got word from somebody else (@xmppwocky) that this *should* work.

**Comment by untitaker** [comment]
Eek, that doesn't even work with &str!

**Comment by untitaker** [comment]
This leaves us with two options:

* `decode::<MyClaims, _>(token, "secret")`
* `decode::<MyClaims>(token, "secret".as_ref())`

For the sake of backwards compat you could also have a dedicated function `decode2` that has either of those interfaces, and make `decode` wrap around it.

**Comment by Keats** [comment]
Ha too bad.
I don't have strong feelings for any of the choice to be honest but i think the second one (`as_ref`) might be easier to figure out what you forgot it from the compiler error so i'd lean towards that.

I will add a changelog section in the readme to document that, no need for backwards compat yet imo

**Comment by untitaker** [comment]
@Keats Updated, please take a look.

File: src/lib.rs [hunk 1]
```diff
@@ -127,7 +127,15 @@ pub fn encode<T: Part, B: AsRef<[u8]>>(claims: &T, secret: B, algorithm: Algorit
...
 /// If the token or its signature is invalid, it will return an error
-pub fn decode<T: Part>(token: &str, secret: &str, algorithm: Algorithm) -> Result<T, Error> {
+pub fn decode<T: Part>(token: &str, secret: &[u8], algorithm: Algorithm) -> Result<T, Error> {
+    // We use &AsRef<[u8]> instead of a separate type parameter such that one still can use:
+    //
+    //     decode::<MyStruct>(...)
+    //
+    // instead of:
+    //
+    //     decode::<MyStruct, _>(...)
```
Comment (Keats) [inline src/lib.rs]: that comment is not valid anymore is it?

Comment (untitaker) [inline src/lib.rs]: Updated.

**Comment by Keats** [comment]
Thanks!

**Pull request merged**